src/cxx_support.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for `cxx` types.
+
+use std::pin::Pin;
+
+use cxx::memory::UniquePtrTarget;
+use cxx::UniquePtr;
+
+use crate::Emplace;
+use crate::TryNew;
+
+/// A type which has the ability to create heap storage space
+/// for itself in C++, without initializing that storage.
+///
+/// # Safety
+///
+/// Implementers must ensure that the pointer returned by
+/// `allocate_uninitialized_cpp_storage` is a valid, non-null,
+/// pointer to a new but uninitialized storage block, and that
+/// such blocks must be freeable using either of these routes:
+///
+/// * before they're initialized, using `free_uninitialized_cpp_storage`
+/// * after they're initialized, using the destructor of
+///   `std::unique_ptr<T, std::default_delete<T>>`
+pub unsafe trait MakeCppStorage
+where
+  Self: Sized,
+{
+  // Allocate heap space for this type in C++ and return a pointer
+  // to that space, but do not initialize that space (i.e. do not
+  // yet call a constructor).
+  //
+  // # Safety
+  //
+  // Callers must ensure that this space is freed using
+  // `free_uninitialized_cpp_storage`, or is converted into
+  // a [`UniquePtr`] such that it can later be freed by
+  // `std::unique_ptr<T, std::default_delete<T>>`.
+  unsafe fn allocate_uninitialized_cpp_storage() -> *mut Self;
+
+  // Free a C++ allocation which has not yet
+  // had a constructor called.
+  //
+  // # Safety
+  //
+  // Callers guarantee that the pointer here was allocated by
+  // `allocate_uninitialized_cpp_storage` and has not been
+  // initialized.
+  unsafe fn free_uninitialized_cpp_storage(ptr: *mut Self);
+}
+
+impl<T: MakeCppStorage + UniquePtrTarget> Emplace<T> for UniquePtr<T> {
+  fn try_emplace<N: TryNew<Output = T>>(n: N) -> Result<Pin<Self>, N::Error> {
+    unsafe {
+      let uninit_ptr = T::allocate_uninitialized_cpp_storage();
+      let uninit = std::mem::transmute(uninit_ptr.as_mut().unwrap());
+      let result = n.try_new(uninit);
+      if let Err(err) = result {
+        T::free_uninitialized_cpp_storage(uninit_ptr);
+        return Err(err);
+      }
+      // TODO - is it possible to remove the need for Emplace to return a pinned
```

Comment:
I think that, in general, we can't do this. `T::emplace` is analogous to `Box::pin` for API purposes... the real issue is that UniquePtr should just provide a function for unpinning itself.

Reply:
I had a crack at this before seeing your comment, and added a commit which attempts to make this work.

Logic in my head is:
* `Pin<Box<T>>` should be the thing which implements a fundamental emplace-y trait. Fundamentally, `Box<T>` is _not_ safe to emplace into; only `Pin<Box<T>>` is safe to emplace into.
* For convenience, `Box<T>` should actually implement some other trait, which calls through to `Pin<Box<T>>::emplace`, so that callers can say `Box::emplace_pinned(...)` or similar.

That's the idea. In the commit I pushed, `emplace_pinned` is also called `emplace` which gets confusing, but this was to avoid breaking compatibility more than necessary. LMK your thoughts.